doc/03-omnifest/02-external.md

```diff
@@ -1,22 +1,117 @@
 # External
 
-## osbuild
+External directives are directives that are implemented externally from `otk`.
+They are meant to be used to provide target-specific behavior and can be used
+to express things where `otk` is not expressive enough.
+
+## Protocol
+
+Directives are small executables that receive JSON and are expected to output
+JSON again in a specific format. Additional keys are not allowed.
+
+When we have an [omnifest](./index.md) that looks like this:
+
+```yaml
+otk.external.name:
+  child:
+    - 1
+  options: "are here"
+```
+
+The external gets called with the following JSON:
+
+```json
+{
+  "tree": {
+    "otk.external.name": {
+      "child": [1],
+      "options": "are here"
+    }
+  }
+}
+```
+
+And is expected to return the following JSON:
+
+```json
+{
+  "tree": {
+    "what": {
+      "you": "want"
+    }
+  }
+}
+```
+
+Which will lead to the following YAML structure:
+
+```yaml
+tree:
+  what:
+    you: "want"
+```
+
+If the external returns `{}`, an empty object, `otk` will assume that there is
+no tree to replace and remove the node instead. This will turn the following
+YAML:
+
+```yaml
+dummy:
+  otk.external.name:
+    options:
+```
+
+Into this YAML structure:
+
+```yaml
+dummy:
+```
+
+## Paths
+
+`otk` will look for external directives in the following paths, stopping when
+it finds the first match:
+
+- A path defined by the `OTK_EXTERNAL_PATH` environment variable.
+- `/usr/local/libexec/otk`
+- `/usr/libexec/otk`
+- `/usr/local/lib/otk`
+- `/usr/lib/otk`
+
+The filename for an external executable is based on the external name. When the
+following directive is encountered: `otk.external.<name>.<argument>` then
+`otk` will try to find an executable called `otk_external_<name>` in the previously
+mentioned search paths.
+
+It will call this executable with `<argument>` as its first argument. Here are
+some examples:
+
+- `otk.external.foo` -> `otk_external_foo`
+- `otk.external.foo.bar_baz` -> `otk_external_foo bar_baz`
+- `otk.external.foo.bar.baz` -> `otk_external_foo bar baz`
```

Comment:
Already I'm looking at this and thinking "is `bar baz` a single argument with a space or two args?"
If it's the former, how do I do two args?
If it's the latter, how do I do args with spaces?

Neither of these questions exist if the specification doesn't allow arguments in the `otk.external` directive's name and subcommands/arguments just become part of the input.

Reply:
Yep, I'm going to drop it, it's confusing.

Reply:
Dropped.

Reply:
Still there?

Reply:
Dropped, I think I only dropped it form the code implementation initially.